src/zarr/codecs/pipeline/batched.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, TypeVar, Iterable
+import numpy as np
+from dataclasses import dataclass
+
+from zarr.config import config
+from zarr.abc.codec import (
+    ByteGetter,
+    ByteSetter,
+    Codec,
+    ArrayArrayCodec,
+    ArrayBytesCodec,
+    ArrayBytesCodecPartialDecodeMixin,
+    ArrayBytesCodecPartialEncodeMixin,
+    BytesBytesCodec,
+)
+from zarr.codecs.pipeline.core import CodecPipeline
+from zarr.common import concurrent_map
+from zarr.indexing import is_total_slice
+
+if TYPE_CHECKING:
+    from typing import List, Optional, Tuple
+    from zarr.common import ArraySpec, BytesLike, SliceSelection
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+def _unzip2(iterable: Iterable[tuple[T, U]]) -> tuple[list[T], list[U]]:
+    out0: list[T] = []
+    out1: list[U] = []
+    for item0, item1 in iterable:
+        out0.append(item0)
+        out1.append(item1)
+    return (out0, out1)
+
+
+def resolve_batched(codec: Codec, chunk_specs: Iterable[ArraySpec]) -> Iterable[ArraySpec]:
+    return [codec.resolve_metadata(chunk_spec) for chunk_spec in chunk_specs]
+
+
+@dataclass(frozen=True)
+class BatchedCodecPipeline(CodecPipeline):
```

Comment:
Since `CodecPipeline` defines `read_batch` and `write_batch`, what is the relationship between those methods and the new functionality offered by this class? I think a clarifying docstring for the class might be useful here, because I don't find the inheritance relationship intuitive (and ditto for the `HybridPipeline` -- I have no intuition for that that one is for :) )